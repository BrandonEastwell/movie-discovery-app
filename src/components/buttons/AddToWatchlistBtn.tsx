"use client"
import WatchlistPopup from "../WatchlistPopup";
import React, {useState} from "react";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faCirclePlus} from "@fortawesome/free-solid-svg-icons";
import ReactDOM from "react-dom";
import { motion } from "framer-motion";
import {useRouter} from "next/navigation";

export default function AddToWatchlistBtn({ movieId, isLoggedIn } : { movieId: number, isLoggedIn: boolean }) {
    const [isVisible, setIsVisible] = useState(false);
    const [cursorPosition, setCursorPosition] = useState({ x: 0, y: 0 });
    const router = useRouter();

    const handleBtnClick = (event: React.MouseEvent) => {
        event.stopPropagation();
        if (isLoggedIn) {
            const { clientX, clientY } = event;
            setCursorPosition({ x: clientX + 25, y: clientY + 25 });
            setIsVisible(!isVisible);
        } else {
            router.push("/auth/login")
        }
    };

    return (
        <>
            <motion.button whileHover={{scale: 1.1}} whileTap={{scale: 0.9}} onClick={(e) => handleBtnClick(e)}
                    className="p-0 bg-transparent max-h-[32px] max-w-[32px] cursor-pointer">
                <FontAwesomeIcon className="text-pearl-white opacity-75" icon={faCirclePlus} size="xl"/>
            </motion.button>
            {isVisible && ReactDOM.createPortal(
                <motion.div animate={{ scale: isVisible ? 1 : 0 }}  onClick={(event) => { event.stopPropagation(); setIsVisible(false); }} className="absolute left-0 top-0 w-screen h-screen bg-transparent z-10">
                    <WatchlistPopup cursorPosition={cursorPosition} movieId={movieId} />
                </motion.div>, document.body
            )}
        </>
    )
}