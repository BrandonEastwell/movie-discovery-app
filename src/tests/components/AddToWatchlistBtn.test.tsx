import {render, screen} from '@testing-library/react';
import AddToWatchlistBtn from "../../components/buttons/AddToWatchlistBtn";
import userEvent from "@testing-library/user-event";

jest.mock("../../components/WatchlistPopup", () => () => (
    <div data-testid="watchlist-popup">Mocked Popup</div>
));

jest.mock('next/navigation', () => ({
    useRouter: () => ({
        push: jest.fn(),
    }),
}));

describe('Add to watchlist button', () => {
    beforeEach(() => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve({}),
            })
        ) as jest.Mock;
    });

    it('should render AddToWatchListBtn', () => {
        const { container } = render(<AddToWatchlistBtn movieId={1} isLoggedIn={true} />);
        expect(container).toMatchSnapshot();
    });

    it('renders button text', async () => {
        const user = userEvent.setup();

        render(<AddToWatchlistBtn movieId={1} isLoggedIn={true}/>);
        const button = screen.getByRole("button");
        await user.click(button);

        expect(screen.getByTestId("watchlist-popup")).toBeInTheDocument()
    });
});