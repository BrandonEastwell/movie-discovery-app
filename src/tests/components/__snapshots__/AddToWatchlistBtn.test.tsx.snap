// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add to watchlist button should render AddToWatchListBtn 1`] = `
<div>
  <button
    class="p-0 bg-transparent max-h-[32px] max-w-[32px] cursor-pointer"
    tabindex="0"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-circle-plus fa-xl text-pearl-white opacity-75"
      data-icon="circle-plus"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM232 344l0-64-64 0c-13.3 0-24-10.7-24-24s10.7-24 24-24l64 0 0-64c0-13.3 10.7-24 24-24s24 10.7 24 24l0 64 64 0c13.3 0 24 10.7 24 24s-10.7 24-24 24l-64 0 0 64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
